File: java/src/zahalto1/HttpRequest.java
package zahalto1;

import java.io.ByteArrayOutputStream;
import java.io.IOException;

public class HttpRequest {

    enum Method {
        GET,
        POST
    }

    enum Command {
        DATA,
        COUNT
    }

    private Method method;
    private Command command;
    private final ByteArrayOutputStream received;

    Method getMethod() {
        return method;
    }

    Command getCommand() {
        return command;
    }

    public ByteArrayOutputStream getReceived() {
        return received;
    }

    public byte[] getData() {
        return received.toByteArray();
    }

    public HttpRequest(ByteArrayOutputStream buffer) {
        this.received = buffer;
    }

    private static final String NEW_LINE = "\r\n";
    private void parseHeader() throws IOException {
        String messageString = new String(received.toByteArray(), "UTF-8");

        String[] lines = messageString.split(NEW_LINE);
        String[] header = lines[0].split(" ");
        if (header.length < 2) {
            error = true;
        }
        if (header[0].equalsIgnoreCase("POST")) {
            method = HttpRequest.Method.POST;
        } else if (header[0].equalsIgnoreCase("GET")) {
            method = HttpRequest.Method.GET;
        }
        if (header[1].equals("/esw/myserver/data")) {
            command = HttpRequest.Command.DATA;
        } else if (header[1].equals("/esw/myserver/count")) {
            command = HttpRequest.Command.COUNT;
        }

        for (int i = 0; i < lines.length; i++) {
            if (lines[i].startsWith("Content-Length")) {
                String[] parts = lines[i].split(": ");
                contentSize = Integer.parseInt(parts[1]);
                break;
            }
        }

        received.reset();
    }

    private boolean error = false;

    public boolean isError() {
        return error;
    }

    private boolean headerParsed = false;

    public boolean isHeaderParsed() {
        return headerParsed;
    }

    private boolean dataParsed = false;

    public boolean isDataParsed() {
        return dataParsed;
    }

    private int readBytes = 0;
    private int dataOffset = 0;
    private int contentSize = 0;
    private byte prev1 = 0x0, prev2 = 0x0;
    private byte curr1 = 0x0, curr2 = 0x0;
    public void addData(byte[] data, int len) throws IOException {
        int index = 0;

        if (!headerParsed) {
            if (index >= len)
                return;
            if (readBytes == 0) {
                prev1 = data[index++];
                received.write(prev1);
                readBytes++;
                if (index >= len)
                    return;
            }
            if (readBytes == 1) {
                prev2 = data[index++];
                received.write(prev2);
                readBytes++;
                if (index >= len)
                    return;
            }
            if (readBytes == 2) {
                curr1 = data[index++];
                received.write(curr1);
                readBytes++;
                if (index >= len)
                    return;
            }

            for (; index < len; index++, readBytes++) {
                curr2 = data[index];
                received.write(curr2);
                if (curr1 == 0x0d && curr2 == 0x0a
                        && prev1 == 0x0d && prev2 == 0x0a) {
                    dataOffset = readBytes + 1;
                    index++;
                    parseHeader();
                    headerParsed = true;
                    break;
                }
                prev1 = prev2;
                prev2 = curr1;
                curr1 = curr2;
            }
        }

        if (headerParsed) {
            received.write(data, index, len - index);
            readBytes += len;
            if (readBytes - dataOffset >= contentSize) {
                dataParsed = true;
            }
        }
    }
}

File: cpp/httprequest.cpp
#include "HttpRequest.h"

#include <iostream>
#include <time.h>

using namespace std;

const char* HttpRequest::HTTP_OK = "200 OK";
const char* HttpRequest::HTTP_NO_CONTENT = "204 No Content";
const char* HttpRequest::HTTP_BAD_REQUEST = "400 Bad Request";
const char* HttpRequest::NEW_LINE = "\r\n";

HttpRequest::HttpRequest(MemoryPool & mp) : memoryPool(mp)
{
    receivedString = memoryPool.getCharBuffer();
}

HttpRequest::~HttpRequest()
{
    memoryPool.returnCharBuffer(receivedString);
}

void HttpRequest::print()
{
}

void HttpRequest::parseHeader()
{
    stringstream ss;
    for (auto it = received.begin(); it != received.end(); it++)
    {
        ss << *it;
    }

    string line, temp, temp2, temp3;
    getline(ss, line);
    istringstream l1(line);

    // method
    getline(l1, temp, ' ');
    if (temp == "POST")
    {
        method = Method::POST;
    }
    else if (temp == "GET")
    {
        method = Method::GET;
    }

    // path
    getline(l1, temp, ' ');
    if (temp == "/esw/myserver/data")
    {
        command = Command::DATA;
    }
    else if (temp == "/esw/myserver/count")
    {
        command = Command::COUNT;
    }

    while (getline(ss, temp, '\n'))
    {
        stringstream l2(temp);
        getline(l2, temp2, ':');
        getline(l2, temp3, ':');
        if (temp2 == "Content-Length")
        {
            contentSize = stoi(temp3.c_str());
            break;
        }
    }

    received.clear();
}

void HttpRequest::addData(unsigned char * data, int len)
{
    int index = 0;

    if (!headerParsed)
    {
        if (index >= len)
            return;
        if (readBytes == 0) 
        {
            prev1 = data[index++];
            received.push_back(prev1);
            readBytes++;
            if (index >= len)
                return;
        }
        if (readBytes == 1) 
        {
            prev2 = data[index++];
            received.push_back(prev2);
            readBytes++;
            if (index >= len)
                return;
        }
        if (readBytes == 2) 
        {
            curr1 = data[index++];
            received.push_back(curr1);
            readBytes++;
            if (index >= len)
                return;
        }

        for (; index < len; index++, readBytes++)
        {
            curr2 = data[index];
            received.push_back(curr2);
            if (curr1 == 0x0d && curr2 == 0x0a
                    && prev1 == 0x0d && prev2 == 0x0a)
            {
                dataOffset = readBytes + 1;
                index++;
                parseHeader();
                headerParsed = true;
                break;   
            }
            prev1 = prev2;
            prev2 = curr1;
            curr1 = curr2;
        }
    }

    if (headerParsed)
    {
        copy(data + index, data + len, back_inserter(*receivedString));
        readBytes += len;
        if (readBytes - dataOffset >= contentSize) 
        {
            dataParsed = true;
        }
    }
}

string HttpRequest::getResponse(const string & code, const string & data)
{
    char buf[100];
    time_t now = time(0);
    struct tm tm = *gmtime(&now);
    strftime(buf, sizeof buf, "%a, %d %b %Y %H:%M:%S %Z", &tm);

    stringstream ss;
    // Status Code
    ss << "HTTP/1.1 " << code << HttpRequest::NEW_LINE;

    // Date
    ss << "Date: " << buf << HttpRequest::NEW_LINE;

    // Content-Type
    ss << "Content-Type: text/plain";
    ss << HttpRequest::NEW_LINE;

    // Content-Length
    ss << "Content-Length: ";
    if (data.size() > 0)
        ss << data.size();
    else
        ss << "0";
    ss << HttpRequest::NEW_LINE;

    // Content
    if (data.size() > 0)
    {
        ss << HttpRequest::NEW_LINE;
        ss << data;
        ss << HttpRequest::NEW_LINE;
    }
    else
    {
        ss << HttpRequest::NEW_LINE;
    }

    return ss.str();
}